cardano-crypto-class/src/Cardano/Crypto/KES/CompactSum.hs

```diff
@@ -461,3 +468,54 @@ instance ( OptimizedKESAlgorithm d
          )
       => FromCBOR (SigKES (CompactSumKES h d)) where
   fromCBOR = decodeSigKES
+
+
+--
+-- Direct ser/deser
+--
+
+instance ( DirectSerialise m (SignKeyKES d)
+         , DirectSerialise m (VerKeyKES d)
+         , MonadST m
+         , KESAlgorithm d
+         ) => DirectSerialise m (SignKeyKES (CompactSumKES h d)) where
+  directSerialise push (SignKeyCompactSumKES sk r vk0 vk1) = do
+    directSerialise push sk
+    mlockedSeedUseAsCPtr r $ \ptr ->
+      push (castPtr ptr) (fromIntegral $ seedSizeKES (Proxy :: Proxy d))
+    directSerialise push vk0
+    directSerialise push vk1
+
+instance ( DirectDeserialise m (SignKeyKES d)
+         , DirectDeserialise m (VerKeyKES d)
+         , MonadST m
+         , KESAlgorithm d
+         ) => DirectDeserialise m (SignKeyKES (CompactSumKES h d)) where
+  directDeserialise pull = do
+    sk <- directDeserialise pull
+
+    r <- mlockedSeedNew
+    mlockedSeedUseAsCPtr r $ \ptr ->
+      pull (castPtr ptr) (fromIntegral $ seedSizeKES (Proxy :: Proxy d))
+
+    vk0 <- directDeserialise pull
+    vk1 <- directDeserialise pull
+
+    return $! SignKeyCompactSumKES sk r vk0 vk1
+
+
+instance (MonadST m, MonadThrow m)
+         => DirectSerialise m (VerKeyKES (CompactSumKES h d)) where
+  directSerialise push (VerKeyCompactSumKES h) =
+    unpackByteStringCStringLen (hashToBytes h) $ \(ptr, len) ->
+      push (castPtr ptr) (fromIntegral len)
+
+instance (MonadST m, MonadThrow m, MonadFail m, HashAlgorithm h)
+         => DirectDeserialise m (VerKeyKES (CompactSumKES h d)) where
+  directDeserialise pull = do
+    let len :: Num a => a
+        len = fromIntegral $ sizeHash (Proxy @h)
+    allocaBytes len $ \ptr -> do
+      pull ptr len
+      bs <- packByteStringCStringLen (ptr, len)
+      maybe (fail "Invalid hash") return $! VerKeyCompactSumKES <$!> hashFromBytes bs
```

Comment:
Why all the strictness? Why extra allocation, why not use `ForeignPtr`?
```suggestion
    fptr <- mallocForeignPtrBytes len
    withForeignPtr fptr $ \ptr -> do
      pull ptr len
      let bs = fromForeignPtr 0 len fptr
      maybe (fail "Invalid hash") (return . VerKeyCompactSumKES) $ hashFromBytes bs
```

We'll need some helpers for ForeignPtrs that work in `MonadST`, but that's totally doable and better than using `allocaBytes` everywhere.